src/ripple/app/misc/SHAMapStoreImp.cpp

```diff
@@ -446,23 +446,17 @@ SHAMapStoreImp::run()
                     ;
             }
 
-            std::string nextArchiveDir =
-                dbRotating_->getWritableBackend()->getName();
             lastRotated_ = validatedSeq;
-            std::shared_ptr<NodeStore::Backend> oldBackend;
-            {
-                std::lock_guard lock (dbRotating_->peekMutex());
-
-                state_db_.setState (SavedState {newBackend->getName(),
-                                                nextArchiveDir, lastRotated_});
-                clearCaches (validatedSeq);
-                oldBackend = dbRotating_->rotateBackends(
-                    std::move(newBackend),
-                    lock);
-            }
-            JLOG(journal_.warn()) << "finished rotation " << validatedSeq;
+            clearCaches(validatedSeq);
 
-            oldBackend->setDeletePath();
+            SavedState savedState;
+            savedState.writableDb = newBackend->getName();
+            savedState.archiveDb = dbRotating_->rotateBackends(
+                std::move(newBackend));
+            savedState.lastRotated = lastRotated_;
+            state_db_.setState(savedState);
```

Comment:
Curious: could we "wrap" this logic inside a `TRANSACTION`?

Reply:
If you mean a SQLIte transaction, no. Because rotateBackends() is not involved with SQLite, but affects the Backend objects that write to the nodestore.

Reply:
@mtrippled Good point, addressed.